hibernate-core/src/main/java/org/hibernate/hql/internal/ast/exec/DeleteExecutor.java

```diff
@@ -74,20 +77,34 @@ public DeleteExecutor(HqlSqlWalker walker, Queryable persister) {
 					final CollectionType cType = (CollectionType) type;
 					final AbstractCollectionPersister cPersister = (AbstractCollectionPersister) metamodel.collectionPersister( cType.getRole() );
 					if ( cPersister.isManyToMany() ) {
-						if ( persister.getIdentifierColumnNames().length > 1
-								&& notSupportingTuplesInSubqueries ) {
+						Type keyType = cPersister.getKeyType();
+						String[] columnNames;
+						if ( keyType.isComponentType() ) {
+							ComponentType componentType = (ComponentType) keyType;
+							List<String> columns = new ArrayList<>( componentType.getPropertyNames().length );
+							for ( String propertyName : componentType.getPropertyNames() ) {
+								Collections.addAll( columns, persister.toColumns( propertyName ) );
+							}
+							columnNames = columns.toArray( new String[columns.size()] );
```

Comment:
In modern JVM, usually `columns.toArray(new String[0])` is faster than `columns.toArray(new String[columns.size()]`. See https://stackoverflow.com/questions/174093/toarraynew-myclass0-or-toarraynew-myclassmylist-size